core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -188,6 +206,85 @@ class TransactionCoordinator(brokerId: Int,
     txnManager.removeTransactionsForPartition(transactionStateTopicPartitionId)
   }
 
+  def handleEndTransaction(transactionalId: String,
+                           pid: Long,
+                           epoch: Short,
+                           command: TransactionResult,
+                           responseCallback: EndTxnCallback): Unit = {
+
+    val errors = validateTransactionalId(transactionalId)
+    if (errors != Errors.NONE)
+      responseCallback(errors)
+    else
+      txnManager.getTransaction(transactionalId) match {
+        case None =>
+          responseCallback(Errors.INVALID_PID_MAPPING)
+        case Some(metadata) =>
+          metadata synchronized {
+            if (metadata.pid != pid)
+              responseCallback(Errors.INVALID_PID_MAPPING)
+            else if (metadata.epoch != epoch)
+              responseCallback(Errors.PRODUCER_FENCED)
+            else metadata.state match {
+              case Ongoing =>
+                commitOrAbort(transactionalId, pid, epoch, command, responseCallback, metadata)
+              case CompleteCommit =>
+                if (command == TransactionResult.COMMIT)
+                  responseCallback(Errors.NONE)
+                else
+                  responseCallback(Errors.INVALID_TXN_STATE)
+              case CompleteAbort =>
+                if (command == TransactionResult.ABORT)
+                  responseCallback(Errors.NONE)
+                else
+                  responseCallback(Errors.INVALID_TXN_STATE)
+              case _ =>
+                responseCallback(Errors.INVALID_TXN_STATE)
+            }
+          }
+      }
+  }
+
+  private def commitOrAbort(transactionalId: String,
+                            pid: Long,
+                            epoch: Short,
+                            command: TransactionResult,
+                            responseCallback: EndTxnCallback,
+                            metadata: TransactionMetadata) = {
+    val nextState = if (command == TransactionResult.COMMIT) PrepareCommit else PrepareAbort
+    val newMetadata = new TransactionMetadata(pid, epoch, metadata.txnTimeoutMs, nextState, metadata.topicPartitions, time.milliseconds())
+    metadata.prepareTransitionTo(nextState)
+
+    def logAppendCallback(errors: Errors): Unit = {
+      if (errors != Errors.NONE)
+      // TODO: What do we do in this case?
```

Comment:
Here are the errors which are possible here: https://github.com/confluentinc/kafka/blob/exactly-once-transactional-coordinator/core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala#L336

I think we should just propagate those back to the client, which will either retry or propagate further back to the user.